### PR TITLE
Issue #115 - Allow availability domain configurable per subnet

### PR DESCRIPTION
**Pull request by syedthameem85**
# Proposed change

### Automated testing

If you're running automated testing for this module, we would love to hear from you, and potentially integrate it to the module standard workflow.

### Manual testing

If no automated testing is run, please ensure that at least the three steps below are passing without any error.

- [ ] Running `terraform apply` on each example provided with this module provisions the intended resource(s) without any errors.
- [ ] Modifying module's *Input Variables* after initial provisioning behaves as intended, i.e: any updateable properties are ameneded without recreation of the resource(s).
- [ ] Running `terraform destroy` on each example provided with this module destroys all the resources created by this module and only the resources created by this module.

## Checklist before submitting your PR

- [ ] My code follows [the style guidelines of this project](../tree/main/docs/codingconventions.adoc)
- [ ] these changes provision new resources:
  - [ ] I have updated the README introduction section (README.adoc)
  - [ ] I have updated the README introduction section (README.md)
- [ ] these changes adds any new variables:
  - [ ] I have updated [docs/terraformoptions.adoc](../tree/main/docs/terraformoptions.adoc) to include each variable
- [ ] I have updated the changelog to include an entry for these changes
- [ ] I have updated all provided examples, including each README file and all applicable code blocks
- [ ] these changes generates no new warnings
- [ ] Any dependent changes have been merged and published in upstream modules

Note: *If you are not an Oracle employee, to contribute to an Oracle-sponsored open-source project, you need to sign the [Oracle Contributor Agreement (OCA)](https://oca.opensource.oracle.com/).*

### CODE REVIEW

File: terraform.tfvars.example [hunk 1]
```diff
@@ -102,4 +102,5 @@ attached_drg_id = null
 #subnets = {
 #  sub1 = {name = "subnet1",cidr_block = "10.0.4.0/24"}
 #  sub2 = {cidr_block="10.0.5.0/24",type="private"}
+#  sub3 = {cidr_block="10.0.6.0/24",availability_domain="Uocm:PHX-AD-1"}
```
Comment (hyder) [inline terraform.tfvars.example:105]: You should pass an AD number and then dynamically look this up. We cannot expect users to know this format.

File: modules/subnet/subnet.tf [hunk 2]
```diff
@@ -10,6 +10,8 @@ resource "oci_core_subnet" "vcn_subnet" {
   cidr_block     = each.value.cidr_block
   compartment_id = var.compartment_id
   vcn_id         = var.vcn_id
+  availability_domain = lookup(each.value, "availability_domain", null)
```
Comment (hyder) [inline modules/subnet/subnet.tf:13]: Please avoid null if possible.

**Review by hyder — CHANGES_REQUESTED** [review]
please check the feedback

File: terraform.tfvars.example (hunk 1 above)
Comment (NehaMehak) [reply terraform.tfvars.example]: Updated the code with AD number instead of AD name

Comment (syedthameem85) [reply terraform.tfvars.example]: Update the code with AD number instead of AD name

File: modules/subnet/subnet.tf (hunk 2 above)
Comment (syedthameem85) [reply modules/subnet/subnet.tf]: null cannot be avoided as it is required to handle availability domain specific subnets as well as regional subnets.
As per documentation (https://registry.terraform.io/providers/oracle/oci/latest/docs/resources/core_subnet), for creating availability domain specific subnet, availability domain is required and for regional subnets, availability domain is required to be skipped which needs to be handled via null.

File: modules/subnet/subnet.tf [hunk 3]
```diff
@@ -5,11 +5,16 @@ locals {
...
+data "oci_identity_availability_domains" "ads" {
+  compartment_id = var.tenancy_ocid
+}
+
 resource "oci_core_subnet" "vcn_subnet" {
   for_each       = var.subnets
   cidr_block     = each.value.cidr_block
   compartment_id = var.compartment_id
   vcn_id         = var.vcn_id
+  availability_domain = lookup(each.value, "availability_domain", null) != null ? data.oci_identity_availability_domains.ads.availability_domains[each.value.availability_domain - 1].name : null
```
Comment (hyder) [inline modules/subnet/subnet.tf:17]: Thinking about this again, my understanding is that we recommend using regional subnets. Is there a reason you need to make it specific?

Also, if you still want to continue with this AD-specific direction, please look at https://github.com/oracle-terraform-modules/terraform-oci-oke/blob/main/module-iam.tf

Comment (syedthameem85) [reply modules/subnet/subnet.tf:17]: This request titled  _**Issue #115 - Allow availability domain configurable per subnet**_  was opened , if Oracle recommendation is to use regional subnet , should we have this request ?.

Comment (hyder) [reply modules/subnet/subnet.tf:17]: It's still a valid use case, I was just wondering whether we should have it or not. I think some users might need it at scale.

File: modules/subnet/subnet.tf (hunk 2 above)
Comment (hyder) [reply modules/subnet/subnet.tf]: In this case, I think you should have a default value.

Comment (syedthameem85) [reply modules/subnet/subnet.tf]: @hyder - Can you suggest what default value are you expecting? The use case here is to handle local ad specific and regional subnets which require null for creating regional subnets. I checked the link https://github.com/oracle-terraform-modules/terraform-oci-oke/blob/main/module-iam.tf and its not applicable for this use case.

File: modules/subnet/subnet.tf (hunk 3 above)
[the remark this answers is not in the document]
Comment (syedthameem85) [reply modules/subnet/subnet.tf:17]: @hyder - I am using a for loop here and the expression is using the value from each iteration of for loop. So, i think its not possible to capture it via local.

**Comment by syedthameem85** [comment]
This PR is the solution to the issue https://github.com/oracle-terraform-modules/terraform-oci-vcn/issues/115 - Allow availability domain configurable per subnet

File: modules/subnet/subnet.tf (hunk 3 above)
[the remark this answers is not in the document]
Comment (hyder) [reply modules/subnet/subnet.tf:17]: see https://github.com/oracle-terraform-modules/terraform-oci-oke/blob/main/module-iam.tf#L11-L21

File: modules/subnet/variables.tf [hunk 4]
```diff
@@ -49,3 +49,8 @@ variable "freeform_tags" {
     module      = "oracle-terraform-modules/vcn/oci"
   }
 }
+
+variable "tenancy_id" {
+  description = "Tenancy OCID"
+  type        = string
```
Comment (hyder) [inline modules/subnet/variables.tf:55]: you can leave internal modules variables without description. see https://github.com/oracle-terraform-modules/terraform-oci-oke/blob/main/modules/iam/variables.tf

File: terraform.tfvars.example [hunk 5]
```diff
@@ -102,4 +106,5 @@ attached_drg_id = null
 #subnets = {
 #  sub1 = {name = "subnet1",cidr_block = "10.0.4.0/24"}
 #  sub2 = {cidr_block="10.0.5.0/24",type="private"}
```
Comment (hyder) [inline terraform.tfvars.example:108]: your default in the code seems to be public if type is not specified? I think the default should be private if not specified.

Comment (syedthameem85) [reply terraform.tfvars.example:108]: @hyder  - This change is not part of this PR. Please note . If needed a separate issue need to be opened to fix this

File: modules/subnet/subnet.tf (hunk 2 above)
Comment (hyder) [reply modules/subnet/subnet.tf]: Most regions have 1 AD. So perhaps AD 1? @Djelibeybi @devoncrouse what do you think?

Comment (syedthameem85) [reply modules/subnet/subnet.tf]: @Djelibeybi  @devoncrouse  - Can you please comment when time permits ?

**Pull request merged**